loadgen/loadgen.cc

```diff
@@ -1196,6 +1075,307 @@ struct LogOutputs {
+        MidOfBoundaries<scenario>(l_bound, u_bound);
+    LogDetail([field = ToStringPerformanceField<scenario>(search_settings)](
+                  AsyncDetail& detail) {
+      detail(
+          "BinarySearchToFindPeakPerformance: The mid point value of bounds: " +
+          field);
+    });
+
+    PerformanceResult pr(IssueQueries<scenario, TestMode::PerformanceOnly>(
+        sut, search_settings, performance_set, sequence_gen));
+    PerformanceSummary perf_summary{sut->Name(), u_bound, std::move(pr)};
+    if (perf_summary.PerfConstraintsMet(&msg)) {
+      SetPerformanceField<scenario>(l_bound, search_settings);
+    } else {
+      SetPerformanceField<scenario>(u_bound, search_settings);
+    }
+  }
+
+  qsl->UnloadSamplesFromRam(performance_set.set);
+
+  return l_bound;
+}
+
+/// \brief Runs the performance mode, templated by scenario.
+template <TestScenario scenario>
+void RunPerformanceMode(SystemUnderTest* sut, QuerySampleLibrary* qsl,
+                        const TestSettingsInternal& settings,
+                        const std::vector<LoadableSampleSet>& loadable_sets,
+                        SequenceGen* sequence_gen) {
+  LogDetail([](AsyncDetail& detail) { detail("Starting performance mode:"); });
+
+  // Use first loadable set as the performance set.
+  const LoadableSampleSet& performance_set = loadable_sets.front();
+  LoadSamplesToRam(qsl, performance_set.set);
+
+  PerformanceResult pr(IssueQueries<scenario, TestMode::PerformanceOnly>(
+      sut, settings, performance_set, sequence_gen));
+
+  sut->ReportLatencyResults(pr.latencies);
+
+  LogSummary(
+      [perf_summary = PerformanceSummary{sut->Name(), settings, std::move(pr)}](
+          AsyncSummary& summary) mutable { perf_summary.Log(summary); });
+
+  qsl->UnloadSamplesFromRam(performance_set.set);
+}
+
+/// \brief Runs the binary search mode, templated by scenario.
+template <TestScenario scenario>
+void FindPeakPerformanceMode(SystemUnderTest* sut, QuerySampleLibrary* qsl,
+                             const TestSettingsInternal& settings,
+                             const std::vector<LoadableSampleSet>& _,
+                             SequenceGen* sequence_gen) {
+  using namespace find_peak_performance;
+
+  LogDetail([](AsyncDetail& detail) {
+    detail("Starting FindPeakPerformance mode:");
+  });
+
+  TestSettingsInternal found_settings =
+      BinarySearchToFindPeakPerformance<scenario>(sut, qsl, sequence_gen,
+                                                  settings);
+
+  // Print-out the peak performance test setting.
+  LogDetail([field = ToStringPerformanceField<scenario>(found_settings)](
+                AsyncDetail& detail) {
+    detail("FindPeakPerformance: Found peak performance field: " + field);
+  });
+
+  // Must re-generate to avoid invalid set of samples by padded values of the
+  // MultiStream scenarios.
+  std::vector<loadgen::LoadableSampleSet> loadable_sets(
+      loadgen::GenerateLoadableSets(qsl, found_settings));
```

Comment:
I see why you ignore the loadable_sets argument now.

We should probably get rid of that argument and have every function call GenerateLoadableSets instead. Feel free to address this in a follow up patch if you prefer.